python/src/spark_rapids_ml/pipeline.py

```diff
@@ -0,0 +1,108 @@
+#
+# Copyright (c) 2025, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import TYPE_CHECKING, Any, List, Optional
+
+from pyspark import keyword_only
+from pyspark.ml.base import Estimator
+from pyspark.ml.base import Transformer as CPUTransformer
+from pyspark.ml.feature import VectorAssembler
+from pyspark.ml.pipeline import Pipeline as SparkPipeline
+from pyspark.ml.pipeline import PipelineModel as SparkPipelineModel
+from pyspark.ml.tuning import CrossValidator as SparkCrossValidator
+from pyspark.sql import DataFrame
+
+from spark_rapids_ml.core import _CumlEstimator
+from spark_rapids_ml.utils import get_logger
+
+if TYPE_CHECKING:
+    from pyspark.ml._typing import ParamMap, PipelineStage
+
+
+class NoOpTransformer(VectorAssembler):
+    @keyword_only
+    def __init__(
+        self,
+        inputCols: Optional[List[str]] = None,
+        outputCol: Optional[str] = None,
+        handleInvalid: str = "error",
+    ):
+        kwargs = self._input_kwargs
+        super().__init__(**kwargs)
+
+    def _transform(self, dataset: DataFrame) -> DataFrame:
+        return dataset
+
+
+class Pipeline(SparkPipeline):
+    @keyword_only
+    def __init__(self, *, stages: Optional[List["PipelineStage"]] = None):
+        kwargs = self._input_kwargs
+        super(Pipeline, self).__init__(**kwargs)
+
+    def _fit(self, dataset: DataFrame) -> "SparkPipelineModel":
+        stages = self.getStages()
+
+        for i in range(len(stages)):
+
+            if not self._isGPUEstimator(stages[i]):
+                continue
+
+            if (
+                i > 0
+                and isinstance(stages[i - 1], VectorAssembler)
+                and self._allScalar(dataset, stages[i - 1].getInputCols())  # type: ignore
```

Comment:
Does this actually work in the general case of arbitrary Pipelines which can implement almost arbitrary DAGs since input columns of a stage can be output of any previous stage or the pipeline input.

Reply:
Yeah. The PR only works for (vec_assembler, gpu_estimator) pairs in the list of stages, and it does not work for other patterns.

Reply:
Regarding (vec_assembler, gpu_estimator) pairs I meant adjacent pairs, so pipeline would look like (..., vec_assembler, gpu_estimator, ...)

Reply:
Example:  vec_assembler,gpu_logistic,cpu_gbdt  where cpu_gpdt also uses output of vec assembler.  so this pipeline (after fit and transform) adds 3 new columns to input df: the output of vec_assembler, logistic prediction and cpu gbdt prediction.   I believe this is a valid pipeline even though it has two fit/transform stages relying on the same input.   There could be another stage that ensembles these two outputs.   The current code would break on this since cpu_gbdt would be getting columnar input and not vector.   

The point is that most generally pipelines can be arbitrary dags with each stage fit/transform adding new columns and arbitrary references to columns in previous stages as inputs.